### PR TITLE
Update atomics (deprecations + adding missing reduction ops)

### PR DESCRIPTION
**Pull request by dalg24**
Implementing the changes discussed in #7449
There was no objections at the developer meeting to proceed with all the deprecations that were proposed.
Fell free to speak up if you disagree with any of it.

Summary:
* Deprecated `atomic_query_version()` (no known usage and unclear how that is useful)
* Deprecated `atomic_assign()` in favor of `atomic_store()`
* Deprecated `atomic_compare_exchange_strong()`
* Deprecated `atomic_{increment, decrement}`
* ~~Deprecated `atomic_{fetch_nand, nand_fetch, nand}`~~ **withdrawn**
(I only feel strongly about the three first deprecation items)
* Added `atomic_{mod,xor,nand,lshift,rshift}` that were missing

### CODE REVIEW

**Comment by dalg24** [comment]
(pure history rewrite, split one commit in two, had accidentally merged the atomic_compare_exchange_strong() deprecation with adding the missing reduction ops)

**Review by lucbv — APPROVED** [review]
This looks reasonable to me, no objection. Just remember to update the atomics documentation for this

File: core/src/Kokkos_Atomics_Desul_Wrapper.hpp [hunk 1]
```diff
@@ -74,7 +76,7 @@ KOKKOS_INLINE_FUNCTION void store_fence()  { desul::atomic_thread_fence(desul::M
 // load/store
 template<class T> KOKKOS_FUNCTION Impl::enable_if_atomic_t<T,    T> atomic_load (T const* ptr)                              { return desul::atomic_load (const_cast<std::remove_volatile_t<T>*>(ptr),      desul::MemoryOrderRelaxed(), KOKKOS_DESUL_MEM_SCOPE); }
 template<class T> KOKKOS_FUNCTION Impl::enable_if_atomic_t<T, void> atomic_store(T* ptr, Impl::not_deduced_atomic_t<T> val) { return desul::atomic_store(const_cast<std::remove_volatile_t<T>*>(ptr), val, desul::MemoryOrderRelaxed(), KOKKOS_DESUL_MEM_SCOPE); }
-template<class T> KOKKOS_FUNCTION Impl::enable_if_atomic_t<T, void> atomic_assign(T* ptr, Impl::not_deduced_atomic_t<T> val) { atomic_store(ptr, val); }
+template<class T> KOKKOS_DEPRECATED_WITH_COMMENT("Use atomic_store() instead!") KOKKOS_FUNCTION Impl::enable_if_atomic_t<T, void> atomic_assign(T* ptr, Impl::not_deduced_atomic_t<T> val) { atomic_store(ptr, val); }
```
[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/Kokkos_Atomics_Desul_Wrapper.hpp:79]: Because I forgot to do so

[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/Kokkos_Atomics_Desul_Wrapper.hpp:80]: Done

**Review by crtrott — APPROVED** [review]

**Review by nliber — APPROVED** [review]

File: core/src/Kokkos_Atomics_Desul_Wrapper.hpp [hunk 2]
```diff
@@ -117,26 +121,26 @@ template<class T> KOKKOS_FUNCTION Impl::enable_if_atomic_t<T, void> atomic_max(T
...
 template<class T> KOKKOS_FUNCTION Impl::enable_if_atomic_t<T, T> atomic_compare_exchange(T* ptr, Impl::not_deduced_atomic_t<T> expected, Impl::not_deduced_atomic_t<T> desired) { return desul::atomic_compare_exchange(const_cast<std::remove_volatile_t<T>*>(ptr), expected, desired, desul::MemoryOrderRelaxed(), KOKKOS_DESUL_MEM_SCOPE); }
-template<class T> KOKKOS_FUNCTION Impl::enable_if_atomic_t<T, bool> atomic_compare_exchange_strong(T* const dest, Impl::not_deduced_atomic_t<T> expected, Impl::not_deduced_atomic_t<T> desired) {
-  T expected_ref = expected;
-  return desul::atomic_compare_exchange_strong(dest, expected_ref, desired,
-                  desul::MemoryOrderRelaxed(), desul::MemoryOrderRelaxed(), KOKKOS_DESUL_MEM_SCOPE);
-}
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_4
+template<class T> KOKKOS_DEPRECATED_WITH_COMMENT("Use atomic_compare_exchange() instead!") KOKKOS_FUNCTION Impl::enable_if_atomic_t<T, bool> atomic_compare_exchange_strong(T* const ptr, Impl::not_deduced_atomic_t<T> expected, Impl::not_deduced_atomic_t<T> desired) { return expected == atomic_compare_excahnge(ptr, expected, desired); }
```
Comment (dalg24) [inline core/src/Kokkos_Atomics_Desul_Wrapper.hpp:142]: Christian pointed out the typo `atomic_compare_excahnge`

Comment (ndellingwood) [reply core/src/Kokkos_Atomics_Desul_Wrapper.hpp:142]: I just put in a PR, this showed up in some nightlies

Comment (ndellingwood) [reply core/src/Kokkos_Atomics_Desul_Wrapper.hpp:142]: https://github.com/kokkos/kokkos/pull/7467

**Pull request merged**